File: leetcode/cpp/0001_twosum.cpp
// O(n)


class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> lookup;
        for (int i = 0; i < nums.size(); ++i) {
            if (lookup.count(target - nums[i])) {
                return {lookup[target - nums[i]], i};
            }
            lookup[nums[i]] = i;
        }
        return {};
    }
};

//或者

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> record;

        for (int i = 0; i < nums.size(); ++i) {
            auto item = record.find(target - nums[i]);
            if (item != record.end()) {
                return {item->second, i};
            }

            record[nums[i]] = i;
        }
        return {};
    }
};

File: leetcode/python/0001_twosum.py
# o(n)

# best solution in dict

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        d = dict()
        for k, v in enumerate(nums):
            if target - v in d:
                return [d[target - v], k]
            d[v] = k
